### ISSUE TITLE
Rinkeby vs Ropsten communication issues

### ISSUE THREAD

**Issue by asemiankevich** [status-im/status-go#434]
# Problem

Currently we have have the behaviour in messaging the way so messaging does not work between networks, thus said messages sent from account on Ropsten will not be delivered to account on Rinkeby.

# Implementation

1. Investigate how current functionality is implemented
2. Investigate if there is something we can fix to let the networks communicate with each other

# Acceptance Criteria

Rinkeby - Ropsten messaging available

# Notes

Per @adambabik I guess what happens is that peers which have different networks get disconnected. We don’t have only Whisper nodes running.

**Comment by divan** [comment]
I'm currently investigating this issue, and here is the answer for the first question so far.

First of all, nodes discovery starts with different bootnodes for different networks.

Second, this behaviour seems to be correct, as nodes use [RLPx protocol](https://github.com/ethereum/devp2p/blob/master/rlpx.md#introduction) for peer-to-peer communication and serving multiple protocols (LES and Whisper currently) over single connection was one of the objectives of this protocol and kinda makes sense for most cases.

During RLPx handshake nodes are checking protocol names/versions for compatibility and [running protocol-specific handshakes](https://github.com/ethereum/go-ethereum/blob/master/p2p/peer.go#L344) for all running protocols.

Even if we change bootnodes and run two instances of `statusd` on different networks with the same bootnodes, handshakes will fail and p2p-connections will not be establieshed, due to above mentioned reason. While `Whisper5`-handshake will return no error, `les`-handshake will return `Genesis block mismatch` [stopping this peer handling completely](https://github.com/ethereum/go-ethereum/blob/master/p2p/peer.go#L205).

Bottom line, this behaviour is a correct in terms of current go-ethereum implementation.

As for the second questions("cross-network communication"), I need more time to investigate, but my understanding is that it's feasible to modify code in that way that nodes will ignore mismatching protocols (i.e. protocols with failed initialization handshake) and keep communicating over successful ones (if node supports `les/63` and `shh/5` protocols, and `les` is failed — just keep running `shh/5` protocol).

Such an approach will require modification of go-ethereum code and probably violates Ethereum security model a little, but can be investigated further.

**Comment by divan** [comment]
Some notes on how to perform testing for this issue.

Start four terminal windows/tabs. In first two, run:

`./statusd -datadir /tmp/dir1 -networkid=3 -shh -ipc`
`./statusd -datadir /tmp/dir2 -networkid=4 -shh -ipc`

(Note, it uses new flags format from https://github.com/status-im/status-go/pull/441). `-shh` enables Whisper protocol for node and `-ips` allows IPC communication to connect via `geth attach` and get JS console.

In third and fourth terminals, run:

`geth attach /tmp/dir1/geth.ipc`
`geth attach /tmp/dir1/geth.ipc`

it will open two JS consoles.

In first one, run:
```js
var topic = "0x5a4ea131"
var symKeyID = shh.newSymKey()
var symKey = shh.getSymKey(symKeyID)
symKey 
```
and copy output from last command into clipboard.

In the second console, run:
```js
var symKey = <paste from clipboard>
var topic = "0x5a4ea131"
var symKeyID = shh.addSymKey(symKey)
var filter = shh.newMessageFilter({ topics: [topic], symKeyID: symKeyID })
```

Now, you should be able to run in the first console:
```js
shh.post({ topic: topic, payload: "0x23123123", ttl: 10, powTime: 1, powTarget: 0.001, symKeyID: symKeyID });
```
and receive (or not receive) in the second console:
```js
shh.getFilterMessages(filter)
```

---

Also, some commands used to monitor `statusd` outgoing connections on MacOS X:

```sh
export STATUS_PID=<PID of running statusd>
watch -n 1 "lsof -p $STATUS_PID -n -P -iUDP | grep '\->' | tr -s ' ' | cut -d' ' -f9 | cut -d'>' -f2 | cut -d':' -f1 | sort -n | uniq"
```

**Comment by adambabik** [comment]
@divan very solid investigation! Really like it!

> Even if we change bootnodes and run two instances of statusd on different networks with the same bootnodes, handshakes will fail and p2p-connections will not be establieshed, due to above mentioned reason. While Whisper5-handshake will return no error, les-handshake will return Genesis block mismatch stopping this peer handling completely.

A potential solution might be to run just whisper nodes? So, we would have regular and whisper-only bootnodes.

> it's feasible to modify code in that way that nodes will ignore mismatching protocols (i.e. protocols with failed initialization handshake) and keep communicating over successful ones

This seems very dangerous and can potentially introduce a lot of issues. I would avoid that... 

Generally, I am not sure what business value would bring us cross-network communication. How would we solve sending ether in a chat in such a case? Is it something our users expect at all? I think they may feel safer if this feature stays disable actually. cc @asemiankevich

**Comment by divan** [comment]
@adambabik thanks! You're right, running just Whisper nodes would do the trick and I actually just tried it (by disabling `les` in the statusd's `main.go` via config) and can confirm that it works — I can receive message sent from node running on Rinkeby on the node running on Ropsten.

But, again, this solution is still quite limited.

**Comment by divan** [comment]
Closing as `wontfix` after discussion on benefits/costs of the solution with core team.
